router/session.py

```diff
@@ -0,0 +1,103 @@
+    session_start_date, session_start_timestamp = get_date(parsed_start_time), get_timestamp(parsed_start_time)
+    current_date, current_timestamp = get_current_date(), get_current_timestamp()
+    repeated_session_start_datetime = datetime(current_date.year, current_date.month, current_date.day, parsed_start_time.hour, parsed_start_time.minute, parsed_start_time.second)
+    if session_start_date <= current_date:
+        if repeat_schedule is None:
+            return session_start_timestamp <= current_timestamp + session_start_buffer_time
+        else:
+            return repeated_session_start_datetime <= current_timestamp + session_start_buffer_time
+    return False
+
+def is_end_time_valid(end_time: str, repeat_schedule: str):
+    """
+    Checks if session end time is valid
+    - If end time is given:
+        - If session end date is greater than or equal to current date:
+            - If session is not repeating, checks if the current timestamp is less than or equal to session end timestamp
+            - If session is repeating, builds a timestamp using current date and session end time and checks that with current timestamp
+    - Else, always returns True
+    """
+    if end_time is not None:
+        parsed_end_time = datetime.strptime(end_time,'%Y-%m-%dT%H:%M:%SZ')
+        session_end_date, session_end_timestamp = get_date(parsed_end_time), get_timestamp(parsed_end_time)
+        current_date, current_timestamp = get_current_date(), get_current_timestamp()
+        repeated_session_end_datetime = datetime(current_date.year, current_date.month, current_date.day, parsed_end_time.hour, parsed_end_time.minute, parsed_end_time.second)
+        if session_end_date >= current_date:
+            if repeat_schedule is not None:
+                return current_timestamp <= repeated_session_end_datetime
+            return current_timestamp <= session_end_timestamp
+        return False
+    return True
+
+
+def is_repeat_schedule_valid(repeat_schedule: str):
+    """
+    Checks if the repeating schedule matches to the current day.
+    """
+    if repeat_schedule is not None:
+        if repeat_schedule["type"] == "weekly":
+            return get_current_datetime().weekday() in repeat_schedule["params"]
+    return True
+
+@router.post("/get-session-data")
+def get_session_data(session_id: str):
+    """
+    API to get details about a session if the session is active.
+    Otherwise, returns False to denote session is not active.
+    """
+    query_params = {'session_id':session_id}
+    response = requests.get(session_db_url, params=query_params)
+    session_data = response.json()
+    if response.status_code == 200:
+        if session_data["is_active"] and is_start_time_valid(session_data["start_time"], session_data["repeat_schedule"]) and is_end_time_valid(session_data["end_time"], session_data["repeat_schedule"]) and is_repeat_schedule_valid(session_data["repeat_schedule"]):
```

Comment:
instead of putting this whole thing in one line, you can format it in a better way
```py
if (
   session_data["is_active"] 
   and is_start_time_valid(session_data["start_time"], session_data["repeat_schedule"]) 
   and is_end_time_valid(session_data["end_time"], session_data["repeat_schedule"]) 
   and is_repeat_schedule_valid(session_data["repeat_schedule"]
)
```

Reply:
How are we sure that keys like `start_time`, `is_active`, `repeat_schedule`, etc are available in `session_data` object?

I understand that it's the responsibility of the DB to provide the full information, but then this backend doesn't know if full information is coming in or not. This backend should have it's own way of handling such scenarios.

If a client is asking this backend for a session details, and turns out because of some bug, all the data is not available for that particular session in the DB. Some keys are missing.
In that case, due to the current code, the backend will throw a `KeyError` exception. It will not handle the error gracefully and our client who is trying to talk to this backend will get stuck.

The backend should gracefully handle this. There are two options according to me,
- Either manually check if these keys are available in `session_data` object or not. And if any required key is missing, send an appropriate HTTP response, like 404 all data in provided session not found, or something like that.
- Second option is to use pydantic models. You can create a pydantic model (or a schema) in models.py for how a session data that is being fetched from the DB should look like. You can then enforce this structure onto the incoming data. If the incoming data doesn't match all the required fields and their types, pydantic will automatically throw an appropriate error for you, even pointing out which field is missing.